Rewrite Idl to unifiy almost all GenClient and GenServer under a functor

mseri opened the pull request:
Do not merge yet. This is a work in progress and I am posting it here for discussion.

I have now found a structure that compiles and allows to unify GenClient and GenServer for the Result.result, lwt and async interfaces. This does not yet work with the `Exn` version of the Client and Server.

Also some more cleanup needs to be done, e.g. pass the RPCMONAD in a Make functor that generates all the modules and unify the interfaces and signatures.

mseri commented:
The latest version should work, at least passes all the tests on my machine. If you review the code you will see that the current state requires some ugly returns and `Obj.magic` (these could be hidden in a safe interface, like for lwt and async, though). 

There is definitely more work needed, but before moving on I'd like to get some feedback and ideas on how to restructure it further to simplify the issues mentioned above (they are clear if you look at the changes needed for the examples to work).

mseri commented:
- I have an idea to deal with it using GADTS but it makes writing the `RPCMONAD` a bit more involute, on the other hand it should become much simpler to use.

- Naming needs to change, it is kind of confusing and obscure at the moment

gaborigloi commented:
I'd be interested in that idea using GADTS, it's great if that makes it easier to use.
Maybe we could implement the exn functors by passing in the function that handles the ok and error cases, and the result type.

jonludlam commented on src/async/rpc_async.ml:
```diff
@@ -1,166 +1,21 @@
+module M = struct
+  include AsyncIdl.ImplM
+  (* TODO: can we avoid Obj.magic here? *)
+  let deferred x : 'a Async.Deferred.t = Obj.magic (AsyncM.unbox x)
```
Yeah, I think this was the line I didn't cross when trying to do this...! Not that I'm saying it's a showstopper because unifying these things certainly reduces a large amount of copy-pasted code.

mseri replied:
I think this can be avoided, I have only to figure out what is the best way

jonludlam commented on tests/rpc/client_lwt_new.ml:
```diff
@@ -125,7 +125,7 @@ let main () =
 
   let funcs = Server.implementation in
-  let rpc = rpc (Rpc_lwt.server funcs) in
+  let rpc r = rpc (Rpc_lwt.server funcs |> Obj.magic) r |> Obj.magic in
```
Although leaking Obj.magic calls into the code that end-users of the library write is a lot more worrying.

mseri replied:
I thin this should be avoidable as well. One of these calls is actually `Rpc_lwt.M.lwt` but I realised it only later

mseri replied:
Anyways I agree with you that, if this requires leaking `Obj.magic`, then is the wrong approach.

mseri commented:
The latest version does not require any `Obj.magic` at all. The only annoyance is that to apply a type-level symplification, the Identity monad needs to be copy/pasted in the libraries that need to generate cliets and servers using it.

From the test and examples updates it seems clear that a new combinator is needed in the `ErrM`, currently I am using `let (>>>=) x f = T.(x |> unbox |> run |> f)` but there is scope for improvement.

Naming has improved a lot in my opinion, although there is quite a lot of room for improvement.
I don't have (yet) a decent way to support the autogeneration of the `Exn` versions using the same functor.

I guess the next steps are:

- cleaning up the naming scheme
- figure out how to deal with the generation of the `Exn` modules

mseri commented:
The interesting aspect is that `(>>>=)` and the need to inline `IdM` are practically "cheating". If the code is written in a monadic style, then everything works fine!

On the other hand, I think it's nice to keep this "hack" around because it makes the library backward compatible with fairly minor changes.

The neat thing of the monadic style is that, you can then swap your base monad with anything you like: a new thing to add debug statements all over the place, a different engine, a test engine, one with state that allows you to revert and re-apply calls on demand, ... whatever. At that point is all completely generic and swappable

jonludlam commented:
Very cool! Will need to spend a bit of time looking over this to understand it!

mseri commented:
I cleaned up the API further, it is getting simpler and a bit more ergonomic.
It is practically impossible to make the current `Exn` modules backward compatible due to the difference in certain types, so I did a cleanup there as well:

- I deleted `GenClientExn`, it was only used in a test, we always use the `GenClientExnRpc`

- I created the `Idl.Exn` module that contains `GenClient` (former `GenClientExnRpc`), `GenServer` (former `GenServerExn`) and ther rest of the helpers. So instead of `open Idl`, `xcp-idl` will have to change only 5 names and do `open Idl.Exn` instead.

- I polished the new API further. It is not the definitive version for sure but it is way simple to use, and requires less combinators (although I have added a few that I was constantly using for testing)

mseri commented on example/example.ml:
```diff
@@ -56,41 +55,48 @@ module MyAPI(R : RPC) = struct
 
 end
 
+module IdM = struct
```
In the next rewrite, in preparation, I am going to make this into a real exception monad with `type 'a t = Value 'a | Exception exn`, and make the tests consistent

mseri commented:
I am much happier with how it looks now. It has become really a swappable engine

mseri commented on src/lib/idl.ml:
```diff
@@ -1,478 +1,690 @@
+
+[@@@warning "-32-34"]
+
+module type RPCTRANSFORMER = functor (M : MONAD) -> sig
```
Yes, It is kind of legacy you are right. I may resuscitate it to simplify further the interface of `Rpc_lwt` and `Rpc_async`

gaborigloi replied:
That would be useful, because we use the async one in `xapi-storage-script`.

mseri commented on src/lib/idl.ml:
```diff
@@ -1,478 +1,690 @@
+
+    val return : 'a -> 'a box
+
+    val get : 'a box -> 'a M.t
```
Well spotted, with the last interface `put` and `get` are the same as `box` and `unbox`. I like the first more, what do you think?

gaborigloi replied:
I agree, they just sound better :)

mseri commented:
@gaborigloi I have updated the interface with the `make_strict` (not sure it is a good name)

mseri commented:
Merged to include the recent fixes and resolve the merge conflicts (doing a rebase was too painful and error prone)

gaborigloi commented:
@jonludlam Should we just merge & test it on a branch, and then refine it later (by removing e.g. the legacy exn stuff that we don't need)?

gaborigloi commented:
We've agreed to merge this.

mseri commented:
After reading https://discuss.ocaml.org/t/reducing-mirageos-image-size-using-ocamlclean/ I double checked the dependencies, and updated the jbuild file of the ppx rewriter. It does not affect the size of the executables though, so it should not have anything to do with the size

mseri commented:
I have also moved to ppx_tools_versioned that should allow us to drop the need of ppxfind and cppo (needs testing with different compilers though)

mseri commented:
I took the last thing back. Unless ppx_deriving is converted to omp, it is not possible to drop cppo and the use of ppxfind

mseri commented:
Ping, any progress on this?

gaborigloi commented:
We wanted to just merge it and create a feature branch with @jonludlam
